endpoints/bookstore-grpc/client/src/main/java/com/google/endpoints/examples/bookstore/BookstoreClient.java

```diff
@@ -0,0 +1,235 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+package com.google.endpoints.examples.bookstore;
+
+import com.google.protobuf.Empty;
+
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ClientInterceptors;
+import io.grpc.ForwardingClientCall;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+
+import org.apache.commons.cli.CommandLine;
+import org.apache.commons.cli.CommandLineParser;
+import org.apache.commons.cli.DefaultParser;
+import org.apache.commons.cli.HelpFormatter;
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.Options;
+import org.apache.commons.cli.ParseException;
+
+/**
+ * A client application which calls the Bookstore API over gRPC.
+ */
+public final class BookstoreClient {
+
+  private static final String defaultBookstore = "localhost:8000";
+
+  public static void main(String[] args) throws Exception {
+    Options options = createOptions();
+    CommandLineParser parser = new DefaultParser();
+    CommandLine line;
+    try {
+      line = parser.parse(options, args);
+    } catch (ParseException e) {
+      System.err.println("Invalid command line: " + e.getMessage());
+      printUsage(options);
+      return;
+    }
+
+    String address = defaultBookstore;
+    String apiKey = null;
+    String authToken = null;
+    String operation = "list";
+
+    if (line.hasOption("bookstore")) {
+      address = line.getOptionValue("bookstore");
+    }
+
+    if (line.hasOption("api_key")) {
+      apiKey = line.getOptionValue("api_key");
+    }
+
+    if (line.hasOption("auth_token")) {
+      authToken = line.getOptionValue("auth_token");
+    }
+
+    if (line.hasOption("operation")) {
+      operation = line.getOptionValue("operation");
+    }
+
+    // Create gRPC stub.
+    BookstoreGrpc.BookstoreBlockingStub bookstore = createBookstoreStub(
+        address, apiKey, authToken);
+
+    if ("list".equals(operation)) {
+      listShelves(bookstore);
+    } else if ("create".equals(operation)) {
+      createShelf(bookstore);
+    } else if ("enumerate".equals(operation)) {
+      enumerate(bookstore);
+    }
+  }
+
+  private static final class Interceptor implements ClientInterceptor {
```

Comment:
Some javadocs would be nice (here and in the main methods - eg `listShelves`, `createShelf`, `enumerate`), for folks reading the code trying to figure out the different components of all this.